README.md
# df_utils
Set of Data Engineering functions which I use a lot in ETL data processing.

It all started with print(df).. then print(df.values.tolist()).. then prettify with tabulate...
After some time and work, I noticed which parts of code I use the most and in every script, thus I made functions.
Fast forward.. after some amount of functions this is a logical step to Python module on PyPI (soon).

I use them as a local module

```python
import sys
sys.path.append(r"/home/igorp/Documents/CODE")
from df_utils import db, du, fi
```

I separate them in 3 different files: 
* ### **db** - for database connectors
  * ***mssqlsrv***(sqlserver, database, **kwargs)
  * ***oracle***(hostname, database, **kwargs)
  * ***postgresql***(hostname, database, **kwargs)
* ### **du** - as dataframe utils
  * ***df_2_mssqlsrv***(df, engine_name, schema_name, table_name, ifexist)
  * ***df_2_sqlite***(df, db_path, table_name)
  * ***get_xlsx_data***(file, sheet)
  * ***df_merged_headers***(cl, delimiter)
  * ***get_xlsx***(fn, **kwargs)
  * ***from_excel_ordinal***(ordinal, _epoch0=datetime(1899, 12, 31))
  * ***df_append_2_xlsx***(df, file_name, sheet_name)
  * ***df_2_xlsx***(df, fn, sn, ac=1, m=0, s=0, sr=0)
  * ***print_df***(df, **kwargs)
  * ***df_dtypes***(df, mode)
  * ***split_df***(df, lines = 1000)
  * ***df_unique***(df, col, id=0)
  * ***clean_df***(res, fillna='')
* ### **fi** - as file and list helpers
  * ***remove_sublist***(main_list, unwanted_list)
  * ***check_sublist***(main_list, sub_list, exception=0)
  * ***list_2_pickle***(src_list, fn)
  * ***pickle_2_list***(fn)
  * ***list_slice***(list, chunks)
  * ***replace_many***(text, dic)
  * ***fuzzy_compare_lists***(source_list, match_list, limit_level, fast=0)
  * ***list_2_str***(source_list)
  * ***num_2_str***(data)
  * ***new_folder***(folder, mode=0)
  * ***get_file_list***(path, extension, mode=0)
  * ***split_by_size***(path, file_list, size)
  * ***cp_multi_2_one***(file_list, dst_path)

### Example
Let say I want to read an xlsx (MS Excel) file and print the content of a worksheet.

```python
import sys
sys.path.append('/home/igorp/Documents/CODE/GitHub') #place your path to the df_utils folder
from df_utils import du

from pathlib import Path

fn = Path(r'/home/igorp/Documents/CODE/MISC/Python/Proba.xlsx') # Place your path to the source xlsx file

# Now, just read xlsx
df = du.get_xlsx_data(fn, 'Milka')

# print raw DataFrame
du.print_df(df, vt=2, e=0)

# print cleaned DataFrame
res = du.clean_df(df)
du.print_df(res, vt=3)
```

### Source data:
![screenshot_20220405-231144](https://user-images.githubusercontent.com/17882375/161851292-3b150ef1-f5bd-4777-83e5-7d0e0aed2146.png)

### Result:
Notice that first dataframe is printed as is, without cleaning, then for comparison, there is another print with cleaned dataframe.

![screenshot_20220405-231858](https://user-images.githubusercontent.com/17882375/161851313-5fb66668-13f5-4cdd-b1e3-09da510b76d6.png)


change_log.md
# CHANGE HISTORY

2021-09-27 12:47:45 - DU | ADDED df_append_2_xlsx
2021-09-15 15:13:16 - LI | ADDED fuzzy_compare_lists
2021-09-10 15:36:38 - LI | ADDED cp_multi_2_one
2021-09-08 14:50:47 - LI | ADDED replace_many
2021-08-04 08:37    - LI | NEW list_slice
2021-07-30 11:05    - LI | ADDED list_slice
2021-07-14 15:40:30 - LI | ADDED list_2_pickle and pickle_2_list
2021-06-09 12:49:37 - DU | ADDED cleand_df
26.5.2021. 13:13:10 - DB | Unified SQL Server engine connection with both server and Windows authentication
                    - DB | Fixed merged headers data type
14.1.2021. 14:08:57 - DB | Add oracle_con - for connecting to Oracle database in Cenosco environment



db.py
"""
## DATAFRAME UTILS
*#A collection of useful functions for work with database connectors*

👔 by Igor Perković

🛠 CREATED: 2020-10-13 08:39:29
📆 CHANGED: 2021-12-09 22:04:03

---
⚙ PREREQUISITES:
📘 Libraries: cx_Oracle, pyodbc, sqlalchemy

"""


def mssqlsrv(sqlserver, database, **kwargs):
    """
    #*MicroSoft SQL Server connection engine for Python.*

    ---
    ### Arguments:
    **Mandatory**
    - sqlserver:    Host IP or Name. Non-default port number goes after coma.
    - database:     Database name

    **Optional** (defaults are in bold)
    - a_mode:       **engine** | conn
    - a_autocommit:   True | **False**

    - a_username:   If left, use Windows auth.
    - a_password:
    """


    # Default values
    #-----------------
    mode      = 'engine'
    ac        = False
    username  = ''
    password  = ''
    c         = ''

    # Get dynamic argument
    for k,v in kwargs.items():
        # Check all arguments
        # ---------------------------
        #print("%s = %s" % (k, v))

        #------------------------
        # if k == 'a_server':
        #     sqlserver = v
        # if k == 'a_database':
        #     database  = v
        if k == 'a_mode':
            mode = v
        if k == 'a_autocommit':
            ac = v
        if k == 'a_username':
            username = v
        if k == 'a_password':
            password = v

    #---------------------------------------------------------------------------
    conn_string_base = "DRIVER={SQL Server};SERVER="+sqlserver+";DATABASE="+database
    #---------------------------------------------------------------------------
    if  (len(username)>0) & (len(password)>0):
        conn_string = conn_string_base+";UID="+username+";PWD="+password
    else:
        conn_string = conn_string_base+";Trusted_Connection=Yes;"

    try:
        if mode == 'conn':
            import pyodbc

            # SQL Server Database Connection
            result = pyodbc.connect(conn_string, autocommit=ac)

            print('📢 DB Connection SUCCESS on:', sqlserver, database)
            print('----------------------------------------------------------------')
            print('CONNECTION mode, AUTOCOMMIT:', ac, '\n')

            c = result.cursor()

        if mode == 'engine':
            from sqlalchemy import create_engine
            import urllib.parse

            # SQLAlchemy

            params = urllib.parse.quote(conn_string)
            result  = create_engine(f"mssql+pyodbc:///?odbc_connect={params}")

            print('📢 DB Connection SUCCESS on:', sqlserver, database)
            print('----------------------------------------------------------------')
            print('ENGINE mode\n')

            conn = result.raw_connection()
            c = conn.cursor()

        c.execute('SELECT @@version')
        res = c.fetchall()
        print(res[0][0])

        return result

    except:
        print('DB Connection ERROR\n')

def oracle(hostname, database, **kwargs):
    """
    #*Oracle connection engine for Python.*

    ---
    ### Arguments:
    **Mandatory**
    - hostname: Host IP or Name. Non-default port number goes after coma.
    - database: Database name

    **Optional** (defaults are in bold)
    - a_mode:       **engine** | conn
    - a_autocommit:   True | **False**

    - a_username:
    - a_password:
    """


    # Default values
    #-----------------
    mode      = 'engine'
    username  = ''
    password  = ''
    db_port   = '1521'

    # Get dynamic argument
    for k,v in kwargs.items():
        # Check all arguments
        # ---------------------------
        #print("%s = %s" % (k, v))

        #------------------------
        if k == 'a_mode':
            mode = v
        if k == 'a_port':
            db_port = v
        if k == 'a_username':
            username = v
        if k == 'a_password':
            password = v

    try:
        if mode == 'conn':
            import cx_Oracle

            # Connetcion string to Oracle database
            # ------------------------------------------------------------------------------------
            conn = cx_Oracle.connect(username, password, hostname+'/'+database, encoding="UTF-8")
            # ------------------------------------------------------------------------------------

            print('Connecting to database...')
            cur = conn.cursor()
            cur.execute ('SELECT * FROM v$version')
            result = cur.fetchone()
            cur.close()

            print('----------------------------------------------------------------------------')
            print('📢 DB Connection SUCCESS on:', database)
            print(result[0])
            print('----------------------------------------------------------------------------\n')

            return conn

        if mode == 'engine':
            from sqlalchemy import create_engine

            oracle_connection_string = 'oracle+cx_oracle://{usr}:{pwd}@{host}:{port}/{db}'

            engine = create_engine(
                oracle_connection_string.format(
                    usr  = username,
                    pwd  = password,
                    host = hostname,
                    port = db_port,
                    db   = database
                )
            )
            return engine
    except:
            print('DB Connection ERROR\n')

def postgresql(hostname, database, **kwargs):
    import psycopg2 as pg

    # Default values
    #-----------------
    mode      = 'engine'
    username  = ''
    password  = ''

    # Get dynamic argument
    for k,v in kwargs.items():
        # Check all arguments
        # ---------------------------
        #print("%s = %s" % (k, v))

        #------------------------
        if k == 'a_mode':
            mode = v
        if k == 'a_username':
            username = v
        if k == 'a_password':
            password = v

    try:
        if mode == 'conn':
            #  Define database connection
            conn = pg.connect(f'host={hostname} dbname={database} user={username} password={password}')
            # Checking database connection
            #-----------------------------------------------------
            cursor = conn.cursor()
            #Executing an SQL function using the execute() method
            cursor.execute("select version()")
            # Fetch a single row using fetchone() method.
            data = cursor.fetchone()

            print('📢 DB Connection SUCCESS on:', data[0])
            print('----------------------------------------------------------------')
            print('CONN mode\n')

            return conn

        if mode == 'engine':
            from sqlalchemy import create_engine
            import pandas as pd

            result = create_engine(f"postgresql+psycopg2://{username}:{password}@{hostname}/{database}")

            print('📢 DB Connection SUCCESS on:', hostname, database)
            print('----------------------------------------------------------------')
            print('ENGINE mode\n')

            df = pd.read_sql('SELECT version()', con=result)
            print(df.values[0][0])
            print('\n')

            return result

    except:
        print('DB Connection ERROR\n')


fi.py
"""
## FILES and LISTS
*#A collection of useful functions for work with files and lists*

👔 by Igor Perković

🛠 CREATED: 2020-10-13 08:39:29
📆 CHANGED: 2021-12-27 13:42:13

---
⚙ PREREQUISITES:
📘 Libraries: pandas, rapidfuzz

"""

# Lists
def remove_sublist(main_list, unwanted_list):
    """
    *#Remove items in sub-list from main list*

    ---
    ### Returns:
    → Reduced main list

    ---
    ### Arguments:
    - main_list    list from which I want to remove..
    - sub_list     ..this istems.
    """
    ul= set(unwanted_list)
    res = [x for x in main_list if x not in ul]
    return res

def check_sublist(main_list, sub_list, exception=0):
    """
    *#Check if list conatins a whole or partial sublist*

    ---
    ### Returns:
    → True or False

    ---
    ### Arguments:
    - main_list    list in which I try to find..
    - sub_list     ..this sub-list
    - exception=0  with exception of n items. #Default is 0 which means all items of sub-list should be in main list to get True as a result.
    """
    sl = set(sub_list)
    res = [x for x in main_list if x in sl]

    if (len(res) == (len(sub_list) - exception)):
        return True
    else:
        return False

def list_2_pickle(src_list, fn):
    import pickle

    print(f'Storing list into: 💾 {fn}')
    try:
        f = open(fn,'wb')
        pickle.dump(src_list,f)
        f.close()
        print('\n ✔Successfully saved: 💾',fn)
    except:
        print(f'❌ ERROR saving list into {fn} file.')

def pickle_2_list(fn):
    """
    *#Read serialized list from pickle to list*

    ---
    ### Returns
    → list

    ---
    ### Arguments:
    - fn (Path) file name

    """

    import pickle
    res = []
    with (open(fn, 'rb')) as pickle_file:
        while True:
            try:
                res = pickle.load(pickle_file)
            except EOFError:
                break
    print(f'✔ Successfully read {fn} file.')
    return res

def list_slice (list, chunks):
    """
    *#Slice list to smaller and equal pieces (exept the last chunk)*

    ---
    ### Returns
    → Nested list of sliced sublist of equal size

    ---
    ### Arguments:
    - list   (list) source list
    - chunks (int)  desired number of items in sublist

    """
    res = [list[x:x+chunks] for x in range(0, len(list), chunks)]
    return res

def replace_many(text, dic):
    """
    *#Multiple replacement in given text from replacement dictionary*

    ---
    ### Returns
    → New text

    ---
    ### Arguments:
    - text (str)        Text which will be changed
    - dic  (dictionary) Replace dictionary

    """
    for i, j in dic.items():
        text = text.replace(i, j)
    return text

def fuzzy_compare_lists(source_list, match_list, limit_level, fast=0):
    """
    *#Compare 2 lists using RapidFuzz library with Levenstein algorithm.*

    ---
    ### Prerequisites:
    pip install rapidfuzz

    ---
    ### Returns:
    → DataFrame

    ---
    ### Arguments:
    - source_list (list[str])   Items that we will comapre
    - match_list  (list[str])   ...with items from this list
    - limit_level (int)         and use n best scores
    - fast        (int)         scorer is fuzz.QRatio

    """
    import pandas as pd
    from rapidfuzz import process, fuzz

    acc = [] #accumulator
    scorer_part = fuzz.WRatio

    if fast:
        scorer_part = fuzz.QRatio

    for e,i in enumerate(source_list):
        print(f'\nOriginal item: {i} | {e}/{len(source_list)}  {str(round(e/len(source_list)*100,2))}%)')
        print('----------------------------------------------------------')

        row = []
        row.append(i)

        res = process.extract(i, match_list, scorer=scorer_part, limit=limit_level)

        for e in range(limit_level):
            row.append(res[e][0])
            row.append(round(res[e][1],2))
            print(str(round(res[e][1],2)), '% |', res[e][0])

        acc.append(row)

    rc = ['Original']
    dc = []
    for r in range(limit_level):
        dc.append(f'Match {r+1}')
        dc.append(f'Score {r+1} [%]')

    final_cols = rc + dc
    report = pd.DataFrame(acc, columns=final_cols)

    return report

def list_2_str(source_list):
    """
    *#Transform list of items in string.*

    ---
    ### Returns:
    → String

    ---
    ### Arguments:
    - source_list (list[str])

    """
    res = str([i for i in source_list]).strip('[]')
    return res

def num_2_str(data):
    """
    *#Transform tuple of numeric items in tuple of strings.*

    ---
    ### Returns:
    → tuple of strings

    ---
    ### Arguments:
    - data (tuple[int])

    """
    acc = []
    for d in data:
        acc.append(str(d))
    return tuple(acc)



# Files

def new_folder(folder, mode=0):
    """
    *#Creates a new folder*

    ---
    ### Returns:
    → New folder

    ---
    ### Arguments:
    - folder (Path)
    - mode   (int)  0 = Silent, 1 = With messages

    """
    try:
        folder.mkdir(parents=True, exist_ok=False)
    except FileExistsError:
        if mode:
            print("Folder is already there")
        else:
            pass
    else:
        if mode:
            print("Folder was created")
        else:
            pass

def get_file_list(path, extension, mode=0):
    """
    *#Gets the file list from given path.*

    ---
    ### Returns
    → File list

    ---
    ### Arguments:
    path      (Path) *Windows path to the source files*
    extension (str)  *File extension (.doc, .xlsx, ...) with wildcards*
    mode      (int)  *0 = full path
                     1 = File names only*

    """
    file_list  = []

    for f in path.glob(f'{extension}'):
        if mode == 1:
            file_list.append(f.name)
        else:
            file_list.append(f)

    return file_list

def split_by_size(path, file_list, size):
    """
    *#Split file list into smaller lists of wanted overall size.*

    ---
    ### Returns:
    → a list with splitted sublists

    ---
    ### Arguments:
    - path      (Path)
    - file_list (list[str])
    - size      (int)       Desired size

    """

    from pathlib import Path
    # Accumulator
    acc = 0

    chunk_list = []
    global_list = []

    for f in file_list:
        fp = Path(path / f)
        file_size = fp.stat().st_size
        file_size_MB = round(file_size/1024/1024,2)
        acc = acc + file_size_MB
        print('Size:',file_size_MB,'MB', '📦', round(acc,2),'MB', '💾', fp.name)

        if acc < size:
            chunk_list.append(f)
            print('✔ Collecting more...')
        else:
            print('❌ Stop collecting ❌')
            chunk_list.append(f) # Get last element after full list
            global_list.append(chunk_list) # Add to global list
            # Reset chunk accumulator and list
            acc = 0
            chunk_list=[]

    return global_list

def cp_multi_2_one(file_list, dst_path):
    """
    *#Copy files from many different folders to one destination folder.*

    ---
    ### Returns:
    → All files in one folder

    ---
    ### Arguments:
    - file_list (list[str])
    - dst_path  (Path)      Destination path

    """

    import shutil
    from shutil import SameFileError
    from pathlib import Path

    cn = 0
    err = []
    print('⚙⚙⚙ Copying in process ⚙⚙⚙')
    for f in file_list:
        src = Path(f)

        try:
            shutil.copy(src, dst_path)
            print('🎯 Not existing file...', f)
            cn = cn + 1
        except SameFileError:
            print("➡ Skipping...Existing file", f)
            err.append(f)
        except:
            print("❌❌❌ Unknown error ❌❌❌", f)
            err.append(f)

    print('⚡⚡⚡  Copying is DONE!  ⚡⚡⚡')
    print(f'\n{cn} files successfully copied!')

    print(f'\n{str(len(err))} These files are not copied:')
    for e in err:
        print(e)


du.py
"""
## DATAFRAME UTILS
*#A collection of useful functions for optimized and faster work with DataFrame*

---
👔 by Igor Perković

🛠 CREATED: 2020-10-13 08:39:29
📆 CHANGED: 2022-04-02 00:15:42

---
⚙ PREREQUISITES:
📘 Libraries: xlsxwriter, pandas, sqlalchemy

"""

import numpy as np
import pandas as pd

from datetime import datetime, timedelta


# Databases
def df_2_mssqlsrv(df, engine_name, schema_name, table_name, ifexist):
    print('Inserting into database...')

    # Optimal chunk-size for SQL Server import
    chunk_size=999//(df.shape[1]+1)

    try:
        df.to_sql(name=table_name, con=engine_name, schema=schema_name, index=False, if_exists=ifexist, chunksize=chunk_size, method='multi')
        print(f'Successfully INSERTED table: {table_name}\n')
    except:
        print(f'INSERT FAILED for {table_name}')

def df_2_sqlite(df, db_path, table_name):
    """
    *#Create a SQLite database file from DataFrame*

    ---
    ### Returns
    → SQLite file (database)

    ---
    ### Arguments:
    - df         (DataFrame)
    - db_path    (Path) Database name with path
    - table_name (str)  Table name

    """
    from sqlalchemy import create_engine

    engine = create_engine(f'sqlite:///{db_path}', echo=False)
    sqlite_connection = engine.connect()

    try:
        df.to_sql(table_name, sqlite_connection, if_exists='fail')
        print(f'Successfully created database {db_path} and table {table_name}')
    except:
        print('❌❌❌ Error creating SQLite database !❌❌❌')



# Excel
def get_xlsx_data(file, sheet):
    try:
        print('Getting the data from xlsx to DataFrame')
        df   = pd.read_excel(file, sheet_name = sheet)
        print('Successfully imported!\n')
        return df
    except:
        print('Oh, no! Data Import ERROR!')

def df_merged_headers(cl, delimiter):
    merged_headers = []
    for c in cl:
        tmp = []
        for i in c:
            if ('Unnamed' in str(i)) or ('*' in str(i)) or (i is np.nan) or ('None' in str(i)):
                pass
            else:
                tmp.append(str(i).replace('\n','').strip().upper())
        merged_headers.append(delimiter.join(tmp[0: len(tmp)]))
    return merged_headers

def get_xlsx(fn, **kwargs):
    # Read raw source data
    # --------------------

    #-----------------
    # Default values
    #-----------------
    sn = []
    nr = 0
    delimiter = '-'
    src_headers = 0
    dest_pick = ''

    # Get dynamic argument
    for k,v in kwargs.items():
        # Check all arguments
        # ---------------------------
        # print("%s = %s" % (k, v))

        # Define smtp parameters
        #------------------------
        if k == 'a_sn_list':
            sn  = v
        if k == 'a_header_rows':
            nr = v
        if k == 'a_delimiter':
            delimiter = v
        if k == 'a_destination':
            dest_pick = v

    if nr > 1:
        src_headers = [i for i in range(nr)]

    xlsx = pd.ExcelFile(fn)
    sheets = xlsx.sheet_names
    print(sheets)

    dfs = []
    if sn:
        for ws in sn:
            print('Reading worksheet:',ws)
            print('headers: ', src_headers)
            df = xlsx.parse(ws, header=src_headers)
            if nr > 1:
                cl = df.columns.tolist()
                df.columns = df_merged_headers(cl, delimiter)
            print('Appending in general list...')
            dfs.append(df)

        res = pd.concat(dfs)
        res.columns = res.columns.str.replace('\n', '')

        res.reset_index(drop=True, inplace=True)
        res.dropna(axis=1, how='all', inplace=True)
        res.dropna(axis=0, how='all', inplace=True)

        if len(dest_pick)>1:
            res.to_pickle(dest_pick)
            print('Created:',dest_pick)

    print('⚡⚡⚡ DONE ⚡⚡⚡')

def from_excel_ordinal(ordinal, _epoch0=datetime(1899, 12, 31)):
    # Convert Excel date shown as serial number into a date string
    if ordinal >= 60:
        ordinal -= 1  # Excel leap year bug, 1900 is not a leap year!
    return (_epoch0 + timedelta(days=ordinal)).replace(microsecond=0)

def df_append_2_xlsx(df, file_name, sheet_name):
    """
    *#Appends data from DataFrame to a new worksheet in existing file.*

    ### Prerequisites:
    pip install openpyxl

    ---
    ### Returns
    → File list

    ---
    ### Arguments:
    df         (DataFrame) *your data*
    file_name  (Path)      *Existing xlsx file*
    sheet_name (str)       *Sheet Name for a new data*
    """

    from openpyxl import load_workbook
    from openpyxl.utils.dataframe import dataframe_to_rows
    from openpyxl.styles import Font,PatternFill

    try:
        wb = load_workbook(file_name)
        writer = pd.ExcelWriter(file_name, engine='openpyxl')
        writer.book = wb
        new_sheet = wb.create_sheet(title=sheet_name)

        # Freeze 1st row
        for r in dataframe_to_rows(df, index=False, header=True):
            new_sheet.append(r)
        new_sheet.freeze_panes = 'A2'

        # Header row style
        fg_style = Font(color='00FFD966', bold=True)
        bg_style = PatternFill("solid", start_color="000d0d0d")

        for y in range(1, new_sheet.max_column+1):
            new_sheet.cell(row=1, column=y).font = fg_style
            new_sheet.cell(row=1, column=y).fill = bg_style

        writer.save()
        print(f'✔ Successfully append sheet: {sheet_name}\n   to file: {file_name}')
    except:
        print(f'❌❌❌ Cannot open or write to file: {file_name}.\nIs it open ?')

def df_2_xlsx(df, fn, sn, ac=1, m=0, s=0, sr=0):
    """
    *#Saves single Dataframe to single Excel xlsx file.*

    ---
    ### Returns:
    → Single xlsx file with one worksheet with dataframe data

    ---
    ### Arguments:
    - df  (DataFrame)
    - fn  (Path(str))  File Name
    - sn  (str)        Sheet Name
    - ac  (int)        0 = Off,    1 = On    (Auto-resize column)
    - m   (int)        0 = Single, 1 = Multi
    - s   (int)        0 = Header, 1 = Table

    """
    import xlsxwriter
    import pandas as pd

    def get_col_widths(dataframe):
        idx_max = max([len(str(s)) for s in dataframe.index.values] + [len(str(dataframe.index.name))])
        return [idx_max] + [max([len(str(s)) for s in dataframe[col].values] + [len(col)]) for col in dataframe.columns]

    writer = pd.ExcelWriter(fn, engine='xlsxwriter')

    # Declare Excel Workbook
    workbook  = writer.book

    # Workbook properties
    workbook.set_properties({
        'author':   'IgorP',
        'company':  'Private',
        'category': 'Report',
    })

    format_dict = {
        'bold'       : True,
        'text_wrap'  : True,
        'valign'     : 'middle',
        'font_color' : '#ffdf3c',
        'fg_color'   : '#232323',
        'border'     : 1
        }


    if m==0:
        if 'DataFrame' in str(type(df)):

            # Create worksheet Data with data from dataframe
            df.to_excel(writer, sheet_name=sn, index=False)

            # Cosmetic - auto-fit columns
            #-----------------------------------------------
            ws = writer.sheets[sn]

            if ac:
                for i, width in enumerate(get_col_widths(df)):
                    ws.set_column(i-1, i-1, width+2)

            # Style
            #------------------------------------
            if s:
                # Table style format
                #--------------------
                cols = []
                for c in (df.columns.values):
                    cols.append({'header': c})

                ws.add_table(0, 0, len(df.index), len(df.columns)-1, {'columns': cols, 'style': 'Table Style Medium 2'})

            else:
                # Header formatting (option instead of Table Style)
                # -------------------------------------------------
                header_format = workbook.add_format(format_dict)

                for col_num, value in enumerate(df.columns.values):
                    ws.write(0, col_num, value, header_format)

            # Freeze 1st row
            ws.freeze_panes(1, 0)

            # Worksheet zoom level
            ws.set_zoom(80)
        else:
            print('This is not a single dataframe for process.')
            print('Exiting...')
            exit()
    else:
        if 'list' in str(type(df)):
            if len(df) == len(sn):
                for d,s in zip(df, sn):
                    d.to_excel(writer, sheet_name=s, index=False)

                    ws = writer.sheets[s]
                    for i, width in enumerate(get_col_widths(d)):
                        ws.set_column(i-1, i-1, width+2)

                    if s:
                        # Table style format
                        #--------------------
                        cols = []

                        for c in (d.columns.values):
                            cols.append({'header': c})

                        ws.add_table(0, 0, len(d.index), len(d.columns)-1, {'columns': cols, 'style': 'Table Style Medium 2'})

                    else:
                        # Header formatting (option instead of Table Style)
                        # -------------------------------------------------
                        header_format = workbook.add_format(format_dict)

                        for col_num, value in enumerate(d.columns.values):
                            ws.write(0, col_num, value, header_format)

                    # Freeze 1st row
                    ws.freeze_panes(1, 0)

                    # Worksheet zoom level
                    ws.set_zoom(80)
            else:
                print('List of DataFrames and list of WorkSheets are not of the same size.')
                print('Exiting...')
                exit()
        else:
            print('This is not a list of dataframes!')
            print('Exiting...')
            exit()
    try:
        writer.save()
        print('\n ✔Successfully saved: 💾',fn)
    except xlsxwriter.exceptions.FileCreateError:
        print('\n\nERROR!!!\nCannot write in opened file.\nCLOSE THE FILE, PLEASE!\n')


def print_df(df, **kwargs):
    """
    Prints a DataFrame.

    ---
    ### Returns:
    → DataFrame information and values

    ---
    ### Arguments:

    🏁 FLAGS:  1 = ON, 0 = OFF

    ---

    df Dataframe         DataFrame
    d  DTypes            int
    dt DTypes tabular    int
    c  Columns           int
    v  Values            int
    vt Values tabular    int
    e  Exit after print  int

    """
    #-----------------
    # Default values
    #-----------------
    a_d  = 1
    a_dt = 0
    a_c  = 1
    a_v  = 1
    a_vt = 0
    a_e  = 1

    # Get dynamic argument
    for k,v in kwargs.items():
        # Check all arguments
        # ---------------------------
        # print("%s = %s" % (k, v))

        # Define smtp parameters
        #------------------------
        if k == 'd':
            a_d  = v
        if k == 'dt':
            a_dt  = v
        if k == 'c':
            a_c  = v
        if k == 'v':
            a_v  = v
        if k == 'vt':
            a_vt  = v
        if k == 'e':
            a_e  = v

        if a_d==1 and a_dt==1:
            a_d  = 0

        if a_v==1 and a_vt>=1:
            a_v = 0


    print('\n📌 DATAFRAME INFO | Rows:', df.shape[0], 'Columns:', df.shape[1])

    if a_c:
        print('🟢 Columns:',df.columns.tolist(),'\n')

    if a_d:
        print (df.dtypes,'\n')

    if a_dt:
        from tabulate import tabulate

        acc = []
        dtypes  = df.dtypes.tolist()
        columns = df.columns.tolist()

        for c,d in zip(columns, dtypes):
            acc.append([c,d])

        dfd = pd.DataFrame(acc, columns=['Column', 'Dtype'])

        print(tabulate(dfd,headers=dfd.columns,  tablefmt='psql',showindex=False),'\n')

    if a_vt:
        from tabulate import tabulate
        if a_vt == 1:
            print(tabulate(df,headers=df.columns.tolist(),  tablefmt='simple',showindex=False),'\n')
        elif a_vt == 2:
            print(tabulate(df,headers=df.columns.tolist(),  tablefmt='psql',showindex=False),'\n')
        elif a_vt == 3:
            print(tabulate(df,headers=df.columns.tolist(), tablefmt='fancy_grid',showindex=False),'\n')

    if a_v:
        print('\n')
        for v in df.values.tolist():
            print(v)
    if a_e:
        exit()

def df_dtypes(df, mode):
    """
    *#Prints particular DataFrame typ4es*

    ---
    ### Returns
    → Text data

    ---
    ### Arguments:
    - df    (DataFrame)
    - mode  (int)
    ```
       0 = print
       1 = NUM column list
       2 = DAT column list
       3 = TXT column list
       4 = ALL columns list
    ```

    """

    tmp = []
    txt = []
    num = []
    dat = []

    dtypes  = df.dtypes.tolist()
    columns = df.columns.tolist()

    if mode == 0:
        for c,d in zip(columns, dtypes):
            print(d,'\t',c)
        exit()


    elif mode == 1:
        for c,d in zip(columns, dtypes):
            if d in ['float64', 'int64', 'int32']:
                print(d,'\t',c)
                num.append(c)
        return num

    elif mode == 2:
        for c,d in zip(columns, dtypes):
            if d in ['datetime64[ns]']:
                print(d,'\t',c)
                dat.append(c)
        return dat

    elif mode == 3:
        for c,d in zip(columns, dtypes):
            if d == 'O':
                print(d,'\t',c)
                txt.append(c)
        return txt

    else:
        for c,d in zip(columns, dtypes):
            tmp.append([c,d])
        return tmp

def split_df(df, lines = 1000):
    split_list = []
    num_chunks = len(df) // lines + 1
    for i in range(num_chunks):
        split_list.append(df[i*lines:(i+1)*lines])
    return split_list

def df_unique(df, col, id=0):
    res = df.loc[:,col].drop_duplicates()
    res.reset_index(drop=True, inplace=True)
    if id:
        res['ID'] = res.index
    return res

def clean_df(res, fillna=''):
    # Remove empty rows and columns
    res.dropna(how='all', axis=0, inplace=True)
    res.dropna(how='all', axis=1, inplace=True)

    # Strip leading and trailing spaces in column names
    res.columns = res.columns.str.strip()
    # Ensure unique column names
    res.columns = [x[1] if x[1] not in res.columns[:x[0]] else f"{x[1]}_{list(res.columns[:x[0]]).count(x[1])}" for x in enumerate(res.columns)]

    # Strip leading and trailing spaces in all columns
    res = res.apply(lambda x: x.str.strip() if x.dtype == "object" else x)
    # Fillna
    if len(fillna):
        res.fillna(fillna,inplace=True)
    # Remove duplicates
    res = res.drop_duplicates()
    # Reindex
    res.reset_index(drop=True, inplace=True)

    return res
